Added XDM and Persepolis

cycool29 opened the pull request:
https://github.com/subhra74/xdm
https://github.com/persepolisdm/persepolis

Tested on 32-bit but should also work on 64-bit.

cycool29 commented:
@Botspot Any ideas?

Botspot commented:
Hmm, the Import App tool does not seem to be working for this PR.

Botspot commented:
```
Downloading ffmpeg-git-armhf-static.tar.xz... 
Download Results:
gid   |stat|avg speed  |path/URI
======+====+===========+=======================================================
2610fd|OK  |   3.7MiB/s|/home/pi/xdm/ffmpeg.tar.xz

Status Legend:
(OK):download completed.
Done
mv: cannot stat './ffmpeg-git-20211009-armhf-static/ffmpeg': No such file or directory
Failed to move ffmpeg binary file to /opt/xdman/xdman/ !
```

Botspot commented:
Persepolis wasn't even installing the local deb.

Botspot commented:
And on my system, there is no package named `adwaita-qt`.

Botspot commented:
Looks like both of these apps need work.

cycool29 commented:
> XDM install failure:
> 
> ```
> Downloading ffmpeg-git-armhf-static.tar.xz... 
> Download Results:
> gid   |stat|avg speed  |path/URI
> ======+====+===========+=======================================================
> 2610fd|OK  |   3.7MiB/s|/home/pi/xdm/ffmpeg.tar.xz
> 
> Status Legend:
> (OK):download completed.
> Done
> mv: cannot stat './ffmpeg-git-20211009-armhf-static/ffmpeg': No such file or directory
> Failed to move ffmpeg binary file to /opt/xdman/xdman/ !
> ```

That's because ffmpeg uses soft-coded link, and the ffmpeg version when I create this PR is not same with current one, hence the link is correct, but not the folder name. Fixed by copying system ffmpeg binary instead of downloading a new one.

> Persepolis wasn't even installing the local deb.

Fixed.

> And on my system, there is no package named adwaita-qt or python3-setproctitle.

Weird. Both of them are in raspbian.raspberrypi.org.
However, since you mentioned it, I have added to download the debs in the install script and install as local deb.

Itai-Nelken commented on apps/Persepolis Download Manager/install:
```diff
@@ -0,0 +1,13 @@
+wget -O adwaita-qt.deb http://mirror.datamossa.io/raspbian/pool/main/a/adwaita-qt/adwaita-qt_1.0-2_armhf.deb
+wget -O python3-setproctitle.deb http://mirror.rise.ph/raspbian/raspbian/pool/main/p/python-setproctitle/python3-setproctitle_1.1.10-1+b2_armhf.deb
+wget -O persepolis.deb https://github.com/persepolisdm/persepolis/releases/download/3.2.0/persepolis_3.2.0.2_all.deb 
+"${DIRECTORY}/pkg-install" "$(pwd)/persepolis.deb $(pwd)/python3-setproctitle.deb $(pwd)/adwaita-qt.deb aria2 python3 sound-theme-freedesktop libnotify-bin python3-pyqt5 libqt5svg5 python3-requests python3-setuptools python3-psutil youtube-dl ffmpeg python3-pyqt5.qtsvg" "$(dirname "$0")" || error "Failed installing Persepolis Download Manager!"
```
It's better to install youtube-dl from `pip`, it's has the newer versions.

cycool29 replied:
Actually I need its binary file, pip doesn't provide that. I think we can download latest one from official site.

cycool29 replied:
Updated.

Botspot commented:
XDM won't run.
```
pi@botspot:~ $ sleep 5;raspi2png
pi@botspot:~ $ "/opt/xdman/xdman"
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.UnsupportedClassVersionError: xdman/Main has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:495)
```

Itai-Nelken commented:
This is the part of the XDM error that matters:
```
xdman/Main has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```
To fix that we either need to install a newer version of java or compile XDM ourselves (if it works with older java). compiling java takes a long time on the pi.

Itai-Nelken commented on apps/Persepolis Download Manager/install:
```diff
@@ -0,0 +1,13 @@
+
+cd $HOME
+wget -O adwaita-qt.deb http://mirror.datamossa.io/raspbian/pool/main/a/adwaita-qt/adwaita-qt_1.0-2_armhf.deb
+wget -O youtube-dl.deb http://ftp.us.debian.org/debian/pool/main/y/youtube-dl/youtube-dl_2021.06.06-1_all.deb
```
I think it's better to install `youtube-dl` from `pip3`.

Botspot replied:
This is the route YouTubuddy takes.
https://github.com/Botspot/youtubuddy/blob/main/gui#L43

cycool29 replied:
> I think it's better to install `youtube-dl` from `pip3`. 


Why? The deb downloaded in the script is latest version of youtube-dl (2021.6.6), same with the [pip one](https://pypi.org/project/youtube_dl/). 

I prefer using deb one because it is `purge-installed`able, and I think using deb (apt) in pi-apps is more reliable than pip since there are more diagnose of apt in pi-apps but not pip. 

Also, if we separate to use pkg-install and pip to install dependencies, that will definetely extend the time to install the app.

Itai-Nelken commented on apps/Xtreme Download Manager/install:
```diff
@@ -0,0 +1,24 @@
+sudo ./install.sh | grep -v "Installation completed" || error "Failed to execute XDM setup script!"
+
+# Use system java
+sudo sed -i s=/opt/xdman/jre/bin/java=java=g /opt/xdman/xdman || error "Failed to change to system Java runtime in /opt/xdman/xdman!"
```
Does this work without this? they probably bundled java because it's the version that works. so unless there is no arm version (probably the case) we should use the included java.

cycool29 replied:
No, using the bundled Java will get `Exec format error` as it is for x64 only. 
Here is where I got the solution: https://github.com/subhra74/xdm/issues/293

cycool29 commented:
> Persepolis and XDM installed fine. Persepolis ran fine but XDM won't run.
> 
> ```
> pi@botspot:~ $ "/opt/xdman/xdman"
> Error: A JNI error has occurred, please check your installation and try again
> Exception in thread "main" java.lang.UnsupportedClassVersionError: xdman/Main has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
> 	at java.lang.ClassLoader.defineClass1(Native Method)
> 	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
> 	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
> 	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
> 	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
> 	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
> 	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
> 	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:495)
> ```

Weird... I haven't met this error. I think error occurs because XDM was compiled with version newer than your jre supports. 

What is your java version `java --version`?

cycool29 commented:
@Botspot Can you try again now?

Crilum commented:
I get this on XDM:
```
OS: Raspbian GNU/Linux 10 (buster)
OS architecture: 32-bit
Pi-Apps local commit ID: a53271dd59a1946cc1eac0acdcefc3405f672f39
Github latest commit ID: a53271dd59a1946cc1eac0acdcefc3405f672f39	
Kernel: armv7l 5.10.63-v7l+
Device model: Raspberry Pi 4 Model B Rev 1.2
Ram size: 3.88 GB
Raspberry Pi OS image version: 2021-05-07
Language: en_US.UTF-8

BEGINNING OF LOG FILE:
-----------------------

Running pkg-install...
Running sudo apt update...


Hit:1 https://itai-nelken.github.io/weekly-box86-debs/debian  InRelease
Hit:4 http://archive.raspberrypi.org/debian buster InRelease
Ign:5 https://download.puffinbrowser.com/repo stable InRelease
Hit:6 https://download.puffinbrowser.com/repo stable Release
Hit:8 http://download.opensuse.org/repositories/home:/stevenpusser/Raspbian_10  InRelease
11 packages can be upgraded. Run 'apt list --upgradable' to see them.
Some packages can be upgraded. Please consider running sudo apt full-upgrade.

Creating dummy deb...

Installing dummy deb...


Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 pi-apps-2db668d5 : Depends: adoptopenjdk-16-hotspot-jre but it is not installable
E: Unable to correct problems, you have held broken packages.

Apt finished.
Failed to install the packages!
APT reported these errors:
E: Unable to correct problems, you have held broken packages.
Failed installing openjdk!

[91mFailed to install Xtreme Download Manager![39m
[40m[93m[5m🔺[25m[39m[49m[93mNeed help? Copy the [1mENTIRE[0m[49m[93m terminal output or take a screenshot.
Please ask on Github: [94m[4mhttps://github.com/Botspot/pi-apps/issues/new/choose[24m[93m
Or on Discord: [94m[4mhttps://discord.gg/RXSTvaUvuu[0m
```

Crilum commented:
Persepolis works.

cycool29 commented:
> archive.raspberrypi.org/debian

Seems `adoptopenjdk-16-hotspot` is not in archive.raspberrypi.org. Added the deb in the script.

cycool29 commented:
Java 11 should work with XDM, so I added a more dynamic way to set how to execute Java:
- If `$(dirname $(dirname $(sudo update-alternatives --list java | grep 'java-11-openjdk' --color=none)))` is found, copy the directory to `/opt/xdman/jre`, without modifying the `/opt/xdman/xdman` file.
- If `java` command exists, change the `/opt/xdman/xdman` to use system Java.

cycool29 commented:
@Botspot Can you test if XDM works now?

Botspot commented:
Both downloaders work now.
